src/big/big_decimal.cr

```diff
@@ -0,0 +1,328 @@
+      case char
+      when '-'
+        if index != 0
+          raise InvalidBigDecimalException.new(str, "Unexpected '-' character")
+        end
+      when '.'
+        if decimal_index
+          raise InvalidBigDecimalException.new(str, "Unexpected '.' character")
+        end
+
+        decimal_index = index
+      when '0'..'9'
+        # Pass
+      else
+        raise InvalidBigDecimalException.new(str, "Unexpected #{char.inspect} character")
+      end
+    end
+
+    if decimal_index
+      value_str = String.build do |builder|
+        # We know this is ASCII, so we can slice by index
+        builder.write(str.to_slice[0, decimal_index])
+        builder.write(str.to_slice[decimal_index + 1, str.bytesize - decimal_index - 1])
+      end
+
+      @value = value_str.to_big_i
+      @scale = (str.bytesize - decimal_index - 1).to_u64
+    else
+      @value = str.to_big_i
+      @scale = 0_u64
+    end
+  end
+
+  # Creates an new `BigDecimal` from `Int`.
+  def initialize(num : Int)
+    initialize(num.to_big_i, 0)
+  end
+
+  # Creating a `BigDecimal` from `Float`.
+  #
+  # NOTE: Floats are fundamentally less precise than BigDecimals, which makes initialization from them risky.
+  def initialize(num : Float)
+    initialize num.to_s
+  end
+
+  # Creates a new `BigDecimal` from `BigInt`/`UInt64`, which matches the internal representation.
+  def initialize(@value : BigInt, @scale : UInt64)
+  end
+
+  def initialize(value : Int, scale : Int)
+    initialize(value.to_big_i, scale.to_u64)
+  end
+
+  def initialize(value : BigInt)
+    initialize(value, 0u64)
+  end
+
+  def initialize
+    initialize(0)
+  end
+
+  # Returns *num*. Useful for generic code that does `T.new(...)` with `T`
+  # being a `Number`.
+  def self.new(num : BigDecimal)
+    num
+  end
+
+  def +(other : BigDecimal) : BigDecimal
+    if @scale > other.scale
+      scaled = other.scale_to(self)
+      BigDecimal.new(@value + scaled.value, @scale)
+    elsif @scale < other.scale
+      scaled = scale_to(other)
+      BigDecimal.new(scaled.value + other.value, other.scale)
+    else
+      BigDecimal.new(@value + other.value, @scale)
+    end
+  end
+
+  def -(other : BigDecimal) : BigDecimal
+    if @scale > other.scale
+      scaled = other.scale_to(self)
+      BigDecimal.new(@value - scaled.value, @scale)
+    elsif @scale < other.scale
+      scaled = scale_to(other)
+      BigDecimal.new(scaled.value - other.value, other.scale)
+    else
+      BigDecimal.new(@value - other.value, @scale)
+    end
+  end
+
+  def *(other : BigDecimal) : BigDecimal
+    BigDecimal.new(@value * other.value, @scale + other.scale)
+  end
+
+  def /(other : BigDecimal) : BigDecimal
+    div other
+  end
+
+  # Divides self with another `BigDecimal`, with a optionally configurable max_div_iterations, which
+  # defines a maximum number of iterations in case the division is not exact.
+  #
+  # ```
+  # BigDecimal(1).div(BigDecimal(2)) => BigDecimal(@value=5, @scale=2)
```

Comment:
Missing `#` before `=>`. On the line below as well.

Reply:
Roger, fixed.